### PR TITLE
[Refactor] Extract common code in gpu and cpu ToBLock

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
This PR refactor GPU `ToBlock` to extract the common code in GPU and GPU to avoid a large redundant code for later changes of CPU `ToBlock ` in PR #5297 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: src/graph/transform/to_block.h [hunk 1]
```diff
@@ -13,23 +13,34 @@
...
 
+#include <functional>
 #include <tuple>
 #include <vector>
 
 namespace dgl {
 namespace transform {
 
+/** @brief Type of the function which maps left and right Id arrays
+ * in a MFG to new ones.
```
Comment (frozenbugs) [inline src/graph/transform/to_block.h:35]: Can you elaborate more here with the format: Mapper for blabla which maps blabla.

File: src/graph/transform/to_block.cc [hunk 2]
```diff
@@ -144,6 +145,174 @@ std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlockCPU(
 
 }  // namespace
 
+template <typename IdType>
+std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlockProcess(
+    HeteroGraphPtr graph, const std::vector<IdArray> &rhs_nodes,
+    bool include_rhs_in_lhs, std::vector<IdArray> *const lhs_nodes_ptr,
+    MappingIdsFunc &&get_maping_ids) {
```
Comment (frozenbugs) [inline src/graph/transform/to_block.cc:152]: IdsMapper &&ids_mapper

File: src/graph/transform/cuda/cuda_to_block.cu [hunk 3]
```diff
@@ -137,209 +137,74 @@ class DeviceNodeMapMaker {
...
+      const std::vector<IdArray>& rhs_nodes,
+      std::vector<IdArray>* const lhs_nodes_ptr,
+      std::vector<int64_t>* const num_nodes_per_type_ptr) {
+    std::vector<IdArray>& lhs_nodes = *lhs_nodes_ptr;
+    std::vector<int64_t>& num_nodes_per_type = *num_nodes_per_type_ptr;
+    const bool generate_lhs_nodes = lhs_nodes.empty();
+    auto device = runtime::DeviceAPI::Get(ctx);
+    cudaStream_t stream = runtime::getCurrentCUDAStream();
+
+    // allocate space for map creation process
```
Comment (frozenbugs) [inline src/graph/transform/cuda/cuda_to_block.cu:156]: Capitalize the first char and end with dot.
For all other comments.

File: src/graph/transform/to_block.cc [hunk 4]
```diff
@@ -13,17 +13,18 @@
  *  See the License for the specific language governing permissions and
  *  limitations under the License.
  *
- * @file graph/transform/to_bipartite.cc
+ * @file graph/transform/to_block.cc
  * @brief Convert a graph to a bipartite-structured graph.
  */
```
Comment (frozenbugs) [inline src/graph/transform/to_block.cc:19]: Add the following snippet here.
```
/**
 * Tested via python wrapper: python/dgl/path/to/to_block.py
 */
```

File: src/graph/transform/to_block.cc [hunk 5]
```diff
@@ -189,6 +358,8 @@ std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlock<kDGLCUDA, int64_t>(
 
 #endif  // DGL_USE_CUDA
 
+// This code is called in `dgl.transform.to_block`, which is tested in
```
Comment (frozenbugs) [inline src/graph/transform/to_block.cc:361]: remove this and apply the snippet on the top.

File: src/graph/transform/to_block.cc (hunk 4 above)
Comment (frozenbugs) [reply src/graph/transform/to_block.cc:19]: If the test of cuda_to_block.cu if also covered by the py test, then add this snippet as well.

**Comment by frozenbugs** [comment]
LGTM overall, please get approval from other reviews to double check the logic.

File: src/graph/transform/to_block.h [hunk 6]
```diff
@@ -49,7 +60,27 @@ std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlock(
     HeteroGraphPtr graph, const std::vector<IdArray>& rhs_nodes,
     bool include_rhs_in_lhs, std::vector<IdArray>* lhs_nodes);
 
+/**
+ * @brief A warpper function shared by CPU and GPU ```ToBlock```
+ * which deal with the common preprocess and postprocess work of them.
+ *
+ * @tparam IdType The type to use as an index.
```
Comment (Rhett-Ying) [inline src/graph/transform/to_block.h:67]: typo `@tparam`?

File: src/graph/transform/to_block.h [hunk 7]
```diff
@@ -13,23 +13,34 @@
...
 
+/** @brief Type of the function which maps left and right Id arrays
+ * in a MFG to new ones.
+ */
+using MappingIdsFunc =
+    std::function<std::tuple<std::vector<IdArray>, std::vector<IdArray>>(
+        const HeteroGraphPtr&, bool, int64_t, const DGLContext&,
+        const std::vector<int64_t>&, const std::vector<EdgeArray>&,
+        const std::vector<IdArray>&, const std::vector<IdArray>&,
+        std::vector<IdArray>* const, std::vector<int64_t>* const)>;
```
Comment (Rhett-Ying) [inline src/graph/transform/to_block.h:42]: why not use reference like previous arguments? legacy? or looks weird to me

File: src/graph/transform/to_block.h [hunk 8]
```diff
@@ -49,7 +60,27 @@ std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlock(
...
+ * @param rhs_nodes The destination nodes of the block.
+ * @param include_rhs_in_lhs Whether or not to include the
+ * destination nodes of the block in the sources nodes.
+ * @param [in/out] lhs_nodes The source nodes of the block.
+ * @param MappingIdsFunc  The function to get mapped ids from original ids.
+ *
+ * @return The block and the induced edges.
+ */
+template <typename IdType>
+std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlockProcess(
```
Comment (Rhett-Ying) [inline src/graph/transform/to_block.h:78]: the function name looks not intuitive to me. come up with a better one?

File: src/graph/transform/cuda/cuda_to_block.cu [hunk 9]
```diff
@@ -137,209 +137,74 @@ class DeviceNodeMapMaker {
...
-    const bool include_rhs_in_lhs, std::vector<IdArray>* const lhs_nodes_ptr) {
-  std::vector<IdArray>& lhs_nodes = *lhs_nodes_ptr;
-  const bool generate_lhs_nodes = lhs_nodes.empty();
-
-  const auto& ctx = graph->Context();
-  auto device = runtime::DeviceAPI::Get(ctx);
-  cudaStream_t stream = runtime::getCurrentCUDAStream();
-
-  CHECK_EQ(ctx.device_type, kDGLCUDA);
-  for (const auto& nodes : rhs_nodes) {
```
Comment (Rhett-Ying) [inline src/graph/transform/cuda/cuda_to_block.cu:154]: These lines seems to be removed now. on purpose?

File: src/graph/transform/to_block.cc [hunk 10]
```diff
@@ -144,6 +145,174 @@ std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlockCPU(
...
+    }
+  }
+
+  std::vector<int64_t> num_nodes_per_type(num_ntypes * 2);
+  // Populate RHS nodes from what we already know.
+  for (int64_t ntype = 0; ntype < num_ntypes; ++ntype) {
+    num_nodes_per_type[num_ntypes + ntype] = rhs_nodes[ntype]->shape[0];
+  }
+
+  std::vector<IdArray> new_lhs;
```
Comment (Rhett-Ying) [inline src/graph/transform/to_block.cc:250]: One picky comment(pls feel free to ignore): this function body is too long. we'd better split it into smaller pieces and we may obtain memory optimization/improvement as some intermediate variables could be destructed/released ASAP.

File: src/graph/transform/to_block.cc [hunk 11]
```diff
@@ -144,6 +145,174 @@ std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlockCPU(
 
 }  // namespace
 
+template <typename IdType>
+std::tuple<HeteroGraphPtr, std::vector<IdArray>> ToBlockProcess(
```
Comment (Rhett-Ying) [inline src/graph/transform/to_block.cc:149]: this will be shared with existing `ToBlockCPU()` in another PR?

**Review by Rhett-Ying — CHANGES_REQUESTED** [review]
generally looks good to me. pls address the comments.

File: src/graph/transform/cuda/cuda_to_block.cu [hunk 12]
```diff
@@ -137,209 +137,74 @@ class DeviceNodeMapMaker {
...
-      edge_arrays[etype] = graph->Edges(etype);
-    }
-  }
-
-  // count lhs and rhs nodes
-  std::vector<int64_t> maxNodesPerType(num_ntypes * 2, 0);
-  for (int64_t ntype = 0; ntype < num_ntypes; ++ntype) {
-    maxNodesPerType[ntype + num_ntypes] += rhs_nodes[ntype]->shape[0];
-
+struct GetMappingIdsGPU {
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply src/graph/transform/cuda/cuda_to_block.cu:141]: FIXED

File: src/graph/transform/to_block.h (hunk 1 above)
Comment (peizhou001) [reply src/graph/transform/to_block.h:35]: Add more description.

File: src/graph/transform/to_block.cc (hunk 2 above)
Comment (peizhou001) [reply src/graph/transform/to_block.cc:152]: Fixed

File: src/graph/transform/cuda/cuda_to_block.cu (hunk 3 above)
Comment (peizhou001) [reply src/graph/transform/cuda/cuda_to_block.cu:156]: fixed

File: src/graph/transform/to_block.cc (hunk 4 above)
Comment (peizhou001) [reply src/graph/transform/to_block.cc:19]: added.

File: src/graph/transform/to_block.cc (hunk 5 above)
Comment (peizhou001) [reply src/graph/transform/to_block.cc:361]: removed

File: src/graph/transform/to_block.h (hunk 7 above)
Comment (peizhou001) [reply src/graph/transform/to_block.h:42]: It is a requirement of google C++ style.

File: src/graph/transform/to_block.h (hunk 8 above)
Comment (peizhou001) [reply src/graph/transform/to_block.h:78]: `ProcessToBLock `looks better?

File: src/graph/transform/cuda/cuda_to_block.cu (hunk 9 above)
Comment (peizhou001) [reply src/graph/transform/cuda/cuda_to_block.cu:154]: It will be executed in another place.

File: src/graph/transform/to_block.cc (hunk 10 above)
Comment (peizhou001) [reply src/graph/transform/to_block.cc:250]: It is directly carried from original `cu_to_block` file. This PR may just move the original code and may not change it.

File: src/graph/transform/to_block.cc (hunk 11 above)
Comment (peizhou001) [reply src/graph/transform/to_block.cc:149]: Yes,  in #5297

**Review by Rhett-Ying — APPROVED** [review]

**Pull request merged**